### PR TITLE
Aarch64 support on Apple platforms

### PR DESCRIPTION
**Pull request by d235j**
I've started working on Aarch64 support on Apple platforms — primarily testing on M1 Mac, but will also be testing on iOS in the future.

This is very much not ready to merge.

Is it worth completing detection for the 32-bit A6/Swift microarchitecture? Apple has discontinued all the devices that use it long ago — perhaps some might want to use asmjit on a jailbroken device, but I don't know if this is worth the extra overhead at all. The only still-shipping Apple platform that uses a 32-bit processor is the Apple Watch Series 3, and it is unlikely that asmjit would be useful (or even usable) there.

### CODE REVIEW

File: src/asmjit/arm/armfeatures.cpp [hunk 1]
```diff
@@ -329,11 +336,139 @@ ASMJIT_FAVOR_SIZE void detectCpu(CpuInfo& cpu) noexcept {
...
 
+#if defined(__APPLE__)
+
+ASMJIT_FAVOR_SIZE void detectCpu(CpuInfo& cpu) noexcept {
+
+  // use sysctlbyname() to obtain CPU family; match on it as there is no good way to detect individual features
+  uint32_t cpuFamily = 0;
+  size_t l = sizeof(cpuFamily);
+  int res;
+  if ((res = sysctlbyname("hw.cpufamily", &cpuFamily, &l, NULL, 0)) != 0) {
```
Comment (kobalicek) [inline src/asmjit/arm/armfeatures.cpp:347]: We have to handle this differently - if we fail to get the cpu family we should assume baseline features only, but not `exit(1)`

File: src/asmjit/core/jitruntime.cpp [hunk 2]
```diff
@@ -45,6 +45,9 @@ static inline void JitRuntime_flushInstructionCache(const void* p, size_t size)
   char* start = static_cast<char*>(const_cast<void*>(p));
   char* end = start + size;
   __builtin___clear_cache(start, end);
+# elif defined(__APPLE__)
+  #include <libkern/OSCacheControl.h>
```
Comment (kobalicek) [inline src/asmjit/core/jitruntime.cpp:49]: The `#include` has to go to the top

File: src/asmjit/arm/armfeatures.cpp (hunk 1 above)
Comment (d235j) [reply src/asmjit/arm/armfeatures.cpp:347]: Resolved for unknown arches, not yet resolved for failed sysctl.

**Comment by d235j** [comment]
Issue: 
In the case of sysctl failure, something has gone really wrong. Not sure what to do.

In the case of a non-matching arch, we currently default to the oldest. This is not necessarily ideal, as future Apple variants will be expected to be supersets of older ones.

Apple says not to use family to match, but they do not provide feature sysctls for all CPU features. I'll probably have to rethink this and/or file feedback with Apple.

**Comment by d235j** [comment]
The `asmjit_test_a64_asm` test has been completed (based on @kobalicek's code from Gitter) and is tested as working on AArch64/M1, with the Builder fix applied.

**Comment by kobalicek** [comment]
@d235j Would you be able to rebase this branch to aarch64? If this is solid I would merge it to aarch64 before I merge aarch64 to master.

**Comment by kobalicek** [comment]
@d235j Could you please rebase it again, and maybe squash it to a single commit? It would be easier for me to review.

Some pointers:
  - we need to get that printf()/exit removed. If we fail to get the cpufamily via the syscall then let's just initialize baseline features
  - FeatureHWDivThumb maps to `kIDIVT`
  - FeatureHWDivARM maps to `kIDIVA`
  - I'm not sure about others - we can just ignore `FeatureHasSlowFPVMLx` for example. Really undecided whether we should report these to users.
  - We should get the remaining flags right - basically these https://github.com/asmjit/asmjit/blob/aarch64/src/asmjit/arm/armfeatures.h#L65 - but if we miss some at the beginning it's not a big deal, we are not able to detect all of them on Linux as well, basically only what OS reports via HW caps.

**Comment by kobalicek** [comment]
I think this PR can be closed - there should be already support for Apple M1 in the current `aarch64` branch.

**Pull request closed without merging**